src/domains/VendingMachine.ts

```diff
@@ -0,0 +1,123 @@
+import Subject from '../core/Subject';
+import { deepClone } from '../utils/commons';
+import { createRandomCoins } from '../utils/coinUtils';
+import {
+  validate,
+  itemValidator,
+  amountValidator,
+  updatedItemValidator,
+  removedItemValidator,
+} from '../utils/validator';
+
+export interface Item {
+  name: string;
+  price: number;
+  quantity: number;
+}
+
+export interface Coins {
+  10: number;
+  50: number;
+  100: number;
+  500: number;
+}
+
+export interface VendingMachineState {
+  items: Item[];
+  coins: Coins;
+}
+
+export default class VendingMachine {
+  state: VendingMachineState;
+
+  constructor(initItems: Item[], initCoins: Coins) {
+    this.init(initItems, initCoins);
+  }
+
+  init(initItems: Item[], initCoins: Coins): void {
+    this.state = Subject.observable({
+      items: initItems,
+      coins: initCoins,
+    });
+  }
+
+  useStore(callback: Function): any {
+    return deepClone(callback(this.state));
+  }
+
+  addItem(item: Item) {
+    const prevItem = this.findItem(item.name);
+
+    if (prevItem) {
+      this.updateItem(prevItem.name, {
+        ...prevItem,
+        ...{ quantity: prevItem.quantity + item.quantity },
+      });
```

Comment:
두 번째 객체는 따로 destructing 안해줘도 되지 않을까요??

```suggestion
      this.updateItem(prevItem.name, {
        ...prevItem,
        quantity: prevItem.quantity + item.quantity,
      });
```

Reply:
그렇네요. suggestion 커밋하겠습니다!